File: src/lib/k8s/port-forward.ts
import { spawn, type ChildProcess } from "child_process";
import { createServer } from "net";
import { prisma } from "@/lib/prisma";

const activeProcesses = new Map<string, ChildProcess>();

interface StartPortForwardOptions {
  contextName: string;
  namespace: string;
  resourceType: string;
  resourceName: string;
  localPort: number;
  remotePort: number;
}

export async function isPortAvailable(port: number): Promise<boolean> {
  return new Promise((resolve) => {
    const server = createServer();
    server.once("error", () => resolve(false));
    server.once("listening", () => {
      server.close(() => resolve(true));
    });
    server.listen(port, "127.0.0.1");
  });
}

export async function startPortForward(opts: StartPortForwardOptions) {
  const { contextName, namespace, resourceType, resourceName, localPort, remotePort } = opts;

  const available = await isPortAvailable(localPort);
  if (!available) {
    throw new Error(`Port ${localPort} is already in use`);
  }

  const session = await prisma.portForwardSession.create({
    data: {
      contextName,
      namespace,
      resourceType,
      resourceName,
      localPort,
      remotePort,
      status: "starting",
    },
  });

  const target = `${resourceType}/${resourceName}`;
  const portMapping = `${localPort}:${remotePort}`;

  const child = spawn(
    "kubectl",
    ["port-forward", target, portMapping, "-n", namespace, "--context", contextName],
    { stdio: ["ignore", "pipe", "pipe"] }
  );

  activeProcesses.set(session.id, child);

  return new Promise<typeof session>((resolve, reject) => {
    let settled = false;
    let stderrOutput = "";

    const timeout = setTimeout(async () => {
      if (!settled) {
        settled = true;
        // If process is still running after 5s without error, assume success
        if (child.exitCode === null) {
          await prisma.portForwardSession.update({
            where: { id: session.id },
            data: { status: "active", pid: child.pid },
          });
          resolve({ ...session, status: "active", pid: child.pid ?? null });
        }
      }
    }, 5000);

    child.stdout?.on("data", async (data: Buffer) => {
      const output = data.toString();
      // kubectl outputs "Forwarding from 127.0.0.1:PORT -> PORT" on success
      if (!settled && output.includes("Forwarding from")) {
        settled = true;
        clearTimeout(timeout);
        await prisma.portForwardSession.update({
          where: { id: session.id },
          data: { status: "active", pid: child.pid },
        });
        resolve({ ...session, status: "active", pid: child.pid ?? null });
      }
    });

    child.stderr?.on("data", (data: Buffer) => {
      stderrOutput += data.toString();
    });

    child.on("error", async (err) => {
      clearTimeout(timeout);
      activeProcesses.delete(session.id);
      await prisma.portForwardSession.update({
        where: { id: session.id },
        data: { status: "error", errorMessage: err.message, stoppedAt: new Date() },
      });
      if (!settled) {
        settled = true;
        reject(new Error(err.message));
      }
    });

    child.on("exit", async (code) => {
      clearTimeout(timeout);
      activeProcesses.delete(session.id);
      const errorMsg = stderrOutput.trim() || (code !== 0 ? `Process exited with code ${code}` : undefined);
      await prisma.portForwardSession.update({
        where: { id: session.id },
        data: {
          status: "stopped",
          errorMessage: errorMsg || undefined,
          stoppedAt: new Date(),
        },
      });
      if (!settled) {
        settled = true;
        if (code !== 0) {
          reject(new Error(errorMsg || `kubectl port-forward exited with code ${code}`));
        } else {
          resolve({ ...session, status: "stopped" });
        }
      }
    });
  });
}

export async function stopPortForward(sessionId: string) {
  const child = activeProcesses.get(sessionId);
  if (child) {
    child.kill("SIGTERM");
    activeProcesses.delete(sessionId);
  }

  await prisma.portForwardSession.update({
    where: { id: sessionId },
    data: { status: "stopped", stoppedAt: new Date() },
  });
}

export async function listActivePortForwards(contextName?: string) {
  const where: Record<string, unknown> = {
    status: { in: ["active", "starting"] },
  };
  if (contextName) {
    where.contextName = contextName;
  }

  const sessions = await prisma.portForwardSession.findMany({
    where,
    orderBy: { createdAt: "desc" },
  });

  // Cross-check with in-memory processes
  for (const session of sessions) {
    if (!activeProcesses.has(session.id)) {
      await prisma.portForwardSession.update({
        where: { id: session.id },
        data: { status: "stopped", stoppedAt: new Date() },
      });
      session.status = "stopped";
    }
  }

  return sessions.filter((s) => s.status === "active" || s.status === "starting");
}

export async function cleanupAllPortForwards() {
  for (const [id, child] of activeProcesses) {
    child.kill("SIGTERM");
    activeProcesses.delete(id);
  }

  await prisma.portForwardSession.updateMany({
    where: { status: { in: ["active", "starting"] } },
    data: { status: "stopped", stoppedAt: new Date() },
  });
}

export async function markStaleSessionsStopped() {
  await prisma.portForwardSession.updateMany({
    where: { status: { in: ["active", "starting"] } },
    data: { status: "stopped", stoppedAt: new Date() },
  });
}

File: backend/src/k8s/port_forward.rs
use anyhow::{anyhow, Result};
use sqlx::SqlitePool;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::process::{Child, Command};
use tokio::sync::Mutex;
use uuid::Uuid;

pub type PortForwardProcesses = Arc<Mutex<HashMap<String, Child>>>;

pub fn new_processes() -> PortForwardProcesses {
    Arc::new(Mutex::new(HashMap::new()))
}

pub async fn is_port_available(port: u16) -> bool {
    tokio::net::TcpListener::bind(("127.0.0.1", port)).await.is_ok()
}

pub async fn start_port_forward(
    db: &SqlitePool,
    processes: &PortForwardProcesses,
    context_name: &str,
    namespace: &str,
    resource_type: &str,
    resource_name: &str,
    local_port: u16,
    remote_port: u16,
) -> Result<serde_json::Value> {
    if !is_port_available(local_port).await {
        return Err(anyhow!("Port {local_port} is already in use"));
    }

    let session_id = Uuid::new_v4().to_string();

    sqlx::query(
        "INSERT INTO port_forward_sessions
         (id, context_name, namespace, resource_type, resource_name, local_port, remote_port, status)
         VALUES (?, ?, ?, ?, ?, ?, ?, 'starting')",
    )
    .bind(&session_id)
    .bind(context_name)
    .bind(namespace)
    .bind(resource_type)
    .bind(resource_name)
    .bind(local_port as i64)
    .bind(remote_port as i64)
    .execute(db)
    .await?;

    let target = format!("{resource_type}/{resource_name}");
    let port_mapping = format!("{local_port}:{remote_port}");

    let mut child = Command::new("kubectl")
        .args([
            "port-forward",
            &target,
            &port_mapping,
            "-n",
            namespace,
            "--context",
            context_name,
        ])
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()?;

    let pid = child.id();

    // Poll for success (wait up to 5s for "Forwarding from" output or process exit)
    let stdout = child.stdout.take();
    drop(stdout); // we won't read it in detail, just wait

    // Give it a moment to start
    tokio::time::sleep(std::time::Duration::from_millis(500)).await;

    // Check process is still alive
    match child.try_wait() {
        Ok(Some(status)) => {
            let code = status.code().unwrap_or(-1);
            sqlx::query(
                "UPDATE port_forward_sessions SET status = 'error',
                 error_message = ?, stopped_at = datetime('now') WHERE id = ?",
            )
            .bind(format!("Process exited with code {code}"))
            .bind(&session_id)
            .execute(db)
            .await?;
            return Err(anyhow!("kubectl port-forward exited immediately with code {code}"));
        }
        Ok(None) => {
            // Still running — mark active
            sqlx::query(
                "UPDATE port_forward_sessions SET status = 'active', pid = ? WHERE id = ?",
            )
            .bind(pid.map(|p| p as i64))
            .bind(&session_id)
            .execute(db)
            .await?;

            processes.lock().await.insert(session_id.clone(), child);
        }
        Err(e) => {
            return Err(anyhow!("Failed to check process status: {e}"));
        }
    }

    let row = sqlx::query_as::<_, PortForwardSession>(
        "SELECT * FROM port_forward_sessions WHERE id = ?",
    )
    .bind(&session_id)
    .fetch_one(db)
    .await?;

    Ok(serde_json::to_value(row)?)
}

pub async fn stop_port_forward(
    db: &SqlitePool,
    processes: &PortForwardProcesses,
    session_id: &str,
) -> Result<()> {
    let mut map = processes.lock().await;
    if let Some(mut child) = map.remove(session_id) {
        let _ = child.kill().await;
    }

    sqlx::query(
        "UPDATE port_forward_sessions SET status = 'stopped', stopped_at = datetime('now')
         WHERE id = ?",
    )
    .bind(session_id)
    .execute(db)
    .await?;

    Ok(())
}

pub async fn list_active_port_forwards(
    db: &SqlitePool,
    processes: &PortForwardProcesses,
    context_name: Option<&str>,
) -> Result<Vec<serde_json::Value>> {
    let rows: Vec<PortForwardSession> = if let Some(ctx) = context_name {
        sqlx::query_as(
            "SELECT * FROM port_forward_sessions WHERE status IN ('active', 'starting') AND context_name = ?
             ORDER BY created_at DESC",
        )
        .bind(ctx)
        .fetch_all(db)
        .await?
    } else {
        sqlx::query_as(
            "SELECT * FROM port_forward_sessions WHERE status IN ('active', 'starting')
             ORDER BY created_at DESC",
        )
        .fetch_all(db)
        .await?
    };

    let map = processes.lock().await;
    let mut result = Vec::new();
    for row in rows {
        if !map.contains_key(&row.id) {
            // Stale — mark stopped
            let _ = sqlx::query(
                "UPDATE port_forward_sessions SET status = 'stopped', stopped_at = datetime('now')
                 WHERE id = ?",
            )
            .bind(&row.id)
            .execute(db)
            .await;
            continue;
        }
        result.push(serde_json::to_value(row)?);
    }

    Ok(result)
}

pub async fn cleanup_all(db: &SqlitePool, processes: &PortForwardProcesses) {
    let mut map = processes.lock().await;
    for (_, mut child) in map.drain() {
        let _ = child.kill().await;
    }

    let _ = sqlx::query(
        "UPDATE port_forward_sessions SET status = 'stopped', stopped_at = datetime('now')
         WHERE status IN ('active', 'starting')",
    )
    .execute(db)
    .await;
}

#[derive(sqlx::FromRow, serde::Serialize)]
pub struct PortForwardSession {
    pub id: String,
    pub context_name: String,
    pub namespace: String,
    pub resource_type: String,
    pub resource_name: String,
    pub local_port: i64,
    pub remote_port: i64,
    pub status: String,
    pub error_message: Option<String>,
    pub pid: Option<i64>,
    pub created_at: String,
    pub stopped_at: Option<String>,
}
